File: backend/app/extractors/mrp.py
"""Extract MRP per Rule 6(1)(e), verifying 'Inclusive of all taxes' is nearby."""

from __future__ import annotations

import re

from app.domain import ExtractedField, ImageMeta, OCRWord
from app.extractors.base import avg_confidence, merge_bboxes

PRICE_PATTERN = re.compile(
    r"(?:(?:M\.?\s*R\.?\s*P\.?|Max(?:imum)?\.?\s*Retail\s*Price)\s*[:\-]?[₹$X\.\s]*"
    r"(?:(?:₹|Rs\.?|Rs|price)\s*[:\-]?[₹$X\.\s]*)?|(?:₹|Rs\.?|Rs|price)\s*[:\-]?[₹$X\.\s]*)"
    r"([0-9,oOlI]+(?:\.[0-9,oOlI]{1,2})?)",
    re.I,
)
MRP_PREFIX_BRANCH = re.compile(
    r"^(?:M\.?\s*R\.?\s*P\.?|Max(?:imum)?\.?\s*Retail\s*Price)\s*[:\-]?",
    re.I,
)


def extract_mrp(
    ocr_words: list[OCRWord],
    image_meta: ImageMeta,
    phrase_regex: str,
    vertical_tolerance: float = 0.06,
) -> ExtractedField | None:
    phrase = re.compile(phrase_regex, re.I)
    candidates: list[dict] = []
    seen_numeric: set[float] = set()

    for i in range(len(ocr_words)):
        sample_text = " ".join(w.text for w in ocr_words[i : min(i + 4, len(ocr_words))])
        if re.search(r"\b(?:save|off|cashback|discount)\b", sample_text, re.I):
            continue
        if re.search(
            r"/\s*(?:g|gm|kg|ml|l|unit|u|pc|piece)\b|\b(?:per\s+(?:g|gm|kg|ml|l|unit|u|pc|piece))\b|\b(?:usp|unit\s*price)\b",
            sample_text,
            re.I,
        ):
            continue

        m = PRICE_PATTERN.match(sample_text)
        if not m and MRP_PREFIX_BRANCH.match(sample_text) and not re.search(r"\d", sample_text):
            sample_text = " ".join(w.text for w in ocr_words[i : min(i + 8, len(ocr_words))])
            m = PRICE_PATTERN.match(sample_text)
        if m:
            raw_val = m.group(1)
            cleaned = (
                raw_val.replace("o", "0").replace("O", "0").replace("l", "1").replace("I", "1")
            )
            if re.search(r"\d", cleaned):
                try:
                    num_val = float(cleaned.replace(",", ""))
                except ValueError:
                    num_val = None
                if num_val is not None and num_val > 0 and num_val not in seen_numeric:
                    seen_numeric.add(num_val)
                    candidates.append({
                        "price_word": ocr_words[i],
                        "value": cleaned,
                        "numeric": num_val,
                    })

    if not candidates:
        return ExtractedField("mrp", None, None, 0.0, [])

    is_normalized = bool(
        ocr_words
        and all(
            w.bbox[0] <= 1.0 and w.bbox[1] <= 1.0 and w.bbox[2] <= 1.0 and w.bbox[3] <= 1.0
            for w in ocr_words
        )
    )
    vert_tol = (
        vertical_tolerance
        if is_normalized
        else vertical_tolerance * (image_meta.height if image_meta.height > 0 else 1000.0)
    )

    best_cand = candidates[0]
    best_phrase_words: list[OCRWord] = []
    best_match = None

    for cand in candidates:
        nearby = [w for w in ocr_words if abs(w.bbox[1] - cand["price_word"].bbox[1]) <= vert_tol]
        joined = " ".join(w.text for w in nearby)
        match = phrase.search(joined)
        if match:
            best_cand = cand
            best_match = match
            pos = 0
            for w in nearby:
                start = pos
                end = start + len(w.text)
                if not (end < match.start() or start > match.end()):
                    best_phrase_words.append(w)
                pos = end + 1
            break

    price_word = best_cand["price_word"]
    value = best_cand["value"]
    conflicting = [c["value"] for c in candidates] if len(candidates) > 1 else []
    extra_spans = [
        c["price_word"].bbox for c in candidates if c["price_word"] is not price_word
    ]

    if not best_match:
        return ExtractedField(
            "mrp",
            None,
            price_word.bbox,
            price_word.confidence,
            [price_word.bbox, *extra_spans],
            conflicting,
        )

    evidence_words = (
        [price_word] + [w for w in best_phrase_words if w is not price_word]
        if best_phrase_words
        else [price_word]
    )
    return ExtractedField(
        "mrp",
        value,
        merge_bboxes(evidence_words),
        avg_confidence(evidence_words),
        [price_word.bbox, *[w.bbox for w in evidence_words if w is not price_word], *extra_spans],
        conflicting,
    )

File: frontend/lib/rules/extractors/mrp.ts
import type { ExtractedField, ImageMeta, OCRWord } from '../domain';
import { avgConfidence, compileRegex, mergeBboxes } from './base';
import { groupWordsIntoSpatialRows } from '../../spatial-layout';

export const PRICE_PATTERN =
  /(?:(?:M\.?\s*R\.?\s*P\.?|Max(?:imum)?\.?\s*Retail\s*Price)\s*[:\-]?[₹$X\.\s]*(?:(?:₹|Rs\.?|Rs|price)\s*[:\-]?[₹$X\.\s]*)?|(?:^|[\s:])(?:₹|Rs\.?|Rs|price)\s*[:\-]?[₹$X\.\s]*)([0-9,oOlI]+(?:\.[0-9,oOlI]{1,2})?)/i;

export const MRP_PREFIX_BRANCH =
  /^(?:M\.?\s*R\.?\s*P\.?|Max(?:imum)?\.?\s*Retail\s*Price)\s*[:\-]?/i;

export const STANDALONE_PRICE =
  /(?:^|[\s:])(?:₹|Rs\.?|INR)?\s*([0-9]{1,5}(?:\.[0-9]{1,2})?)(?!\d)/i;

export const BARCODE_PATTERN = /\b\d{8,}\b/;

export function extractMrp(
  ocrWords: OCRWord[],
  imageMeta: ImageMeta,
  phraseRegex: string,
  verticalTolerance = 0.06
): ExtractedField {
  const phrase = compileRegex(phraseRegex, 'i');
  const candidates: Array<{
    priceWord: OCRWord;
    value: string;
    numeric: number;
    evidenceWords?: OCRWord[];
    columnLeft?: number;
  }> = [];
  const seenNumeric = new Set<number>();

  // Pass 1: Standard sequential window search
  for (let i = 0; i < ocrWords.length; i++) {
    let sampleText = ocrWords
      .slice(i, Math.min(i + 4, ocrWords.length))
      .map((w) => w.text)
      .join(' ');

    if (/\b(?:save|off|cashback|discount)\b/i.test(sampleText)) {
      continue;
    }
    if (
      /\/\s*(?:g|gm|kg|ml|l|unit|u|pc|piece)\b|\b(?:per\s+(?:g|gm|kg|ml|l|unit|u|pc|piece))\b|\b(?:usp|unit\s*price)\b/i.test(
        sampleText
      )
    ) {
      continue;
    }
    if (BARCODE_PATTERN.test(sampleText)) {
      continue;
    }

    let m = PRICE_PATTERN.exec(sampleText);
    if (!m && MRP_PREFIX_BRANCH.test(sampleText) && !/\d/.test(sampleText)) {
      sampleText = ocrWords
        .slice(i, Math.min(i + 8, ocrWords.length))
        .map((w) => w.text)
        .join(' ');
      m = PRICE_PATTERN.exec(sampleText);
    }

    if (m) {
      const rawVal = m[1];
      const cleaned = rawVal
        .replace(/o/g, '0')
        .replace(/O/g, '0')
        .replace(/l/g, '1')
        .replace(/I/g, '1');

      if (/\d{8,}/.test(cleaned)) {
        continue;
      }

      if (/\d/.test(cleaned)) {
        const numVal = parseFloat(cleaned.replace(/,/g, ''));
        if (!isNaN(numVal) && numVal > 0 && numVal < 100000 && !seenNumeric.has(numVal)) {
          seenNumeric.add(numVal);
          candidates.push({
            priceWord: ocrWords[i],
            value: cleaned,
            numeric: numVal,
          });
        }
      }
    }
  }

  // Pass 2: Spatial Row Search for Two-Column Layouts (Left: "M.R.P. ₹", Right: "20.00" / "100.00")
  if (candidates.length === 0) {
    const spatialRows = groupWordsIntoSpatialRows(ocrWords);
    for (const row of spatialRows) {
      if (/m\.?r\.?p\.?|max(?:imum)?\s*retail\s*price/i.test(row.text)) {
        console.log(`[extractMrp] Checking MRP spatial row: "${row.text}"`);
        // Find price tokens in this row
        for (let wIdx = 0; wIdx < row.words.length; wIdx++) {
          const w = row.words[wIdx];
          const anchor = row.words.find((word) => /m\.?r\.?p\.?|maximum|max\.?/i.test(word.text));
          // A manufacturer address in the neighbouring column is not an MRP value.
          if (anchor && w.bbox[0] < anchor.bbox[0]) continue;
          // Skip tokens that are unit prices (e.g. 0.61/g)
          const isUnitPrice = /\/\s*(?:g|gm|kg|ml|l|unit|u|pc)\b|\bper\b|\busp\b/i.test(
            row.words.slice(Math.max(0, wIdx - 1), wIdx + 3).map((item) => item.text).join(' ')
          );
          if (isUnitPrice) continue;

          // Check if word itself or word sequence is a price
          const priceMatch = STANDALONE_PRICE.exec(w.text.replace(/^[₹$]/, ''));
          if (priceMatch) {
            const cleanNum = priceMatch[1].replace(/o/gi, '0').replace(/[lI]/g, '1');
            const num = parseFloat(cleanNum);
            if (!isNaN(num) && num > 0 && num < 100000 && !seenNumeric.has(num)) {
              seenNumeric.add(num);
              candidates.push({
                priceWord: w,
                value: cleanNum,
                numeric: num,
                evidenceWords: row.words,
                columnLeft: anchor ? anchor.bbox[0] - anchor.bbox[2] : undefined,
              });
              console.log(`[extractMrp] Found spatial candidate: ${cleanNum} in row "${row.text}"`);
            }
          }
        }
      }
    }
  }

  if (candidates.length === 0) {
    console.log('[extractMrp] No MRP candidates found in evidence.');
    return {
      name: 'mrp',
      value: null,
      bbox: null,
      confidence: 0.0,
      evidence_spans: [],
      conflicting_values: [],
    };
  }

  const isNormalized =
    ocrWords.length > 0 &&
    ocrWords.every(
      (w) =>
        w.bbox[0] <= 1.0 &&
        w.bbox[1] <= 1.0 &&
        w.bbox[2] <= 1.0 &&
        w.bbox[3] <= 1.0
    );

  const vertTol = isNormalized
    ? verticalTolerance
    : verticalTolerance * (imageMeta.height > 0 ? imageMeta.height : 1000.0);

  let bestCand = candidates[0];
  let bestPhraseWords: OCRWord[] = [];
  let bestMatch: RegExpExecArray | null = null;

  for (const cand of candidates) {
    const nearby = ocrWords.filter(
      (w) => Math.abs(w.bbox[1] - cand.priceWord.bbox[1]) <= vertTol &&
        (cand.columnLeft === undefined || w.bbox[0] >= cand.columnLeft)
    );
    const joined = nearby.map((w) => w.text).join(' ');
    const match = phrase.exec(joined);
    if (match) {
      bestCand = cand;
      bestMatch = match;
      let pos = 0;
      const phraseWords: OCRWord[] = [];
      const matchStart = match.index;
      const matchEnd = match.index + match[0].length;
      for (const w of nearby) {
        const start = pos;
        const end = start + w.text.length;
        if (!(end < matchStart || start > matchEnd)) {
          phraseWords.push(w);
        }
        pos = end + 1;
      }
      bestPhraseWords = phraseWords;
      break;
    }
  }

  const priceWord = bestCand.priceWord;
  const value = bestCand.value;
  const conflicting =
    candidates.length > 1 ? candidates.map((c) => c.value) : [];
  const extraSpans = candidates
    .filter((c) => c.priceWord !== priceWord)
    .map((c) => c.priceWord.bbox);

  if (!bestMatch) {
    return {
      name: 'mrp',
      value: null,
      bbox: priceWord.bbox,
      confidence: priceWord.confidence,
      evidence_spans: [priceWord.bbox, ...extraSpans],
      conflicting_values: conflicting,
    };
  }

  const evidenceWords =
    bestPhraseWords.length > 0
      ? [priceWord, ...bestPhraseWords.filter((w) => w !== priceWord)]
      : [priceWord];

  return {
    name: 'mrp',
    value,
    bbox: mergeBboxes(evidenceWords),
    confidence: avgConfidence(evidenceWords),
    evidence_spans: [
      priceWord.bbox,
      ...evidenceWords.filter((w) => w !== priceWord).map((w) => w.bbox),
      ...extraSpans,
    ],
    conflicting_values: conflicting,
  };
}
